README.md
#DiSCuS
Database System for Compound Selection


##1. System requirements:

###1.1. Server side
* Openbabel + PHP/Python Bindings
* Apache (or any other popular webserver)
* PHP
* Python
* MySQL (or MariaDB, Percona)
* MyChem database cartridge

###1.2. Client Side:
* Modern web browser: Chrome, Firefox, IE 10+ might work, but it's not tested on regular basis

##2. Installation

###2.1 Debian Wheezy

2.1.1. Install dependencies

> apt-get install php5 php5-dev php5-xcache apache2 phpmyadmin mysql-server libmysqld-dev  libmysqlclient-dev git swig cmake

2.1.2. (optional) bonjour to resolve host names 

> apt-get -y install avahi-daemon libnss-mdns

2.1.3. Install dependencies for OB compilation

> apt-get -y build-dep openbabel

2.1.4. Download Openbabel and Mychem (using discus-deploy repository)

> git clone --recursive https://github.com/mwojcikowski/discus-deploy.git discus-deploy

2.1.5. Compile and global install Openbabel and Mychem

> cd discus-deploy

> ./compile_ob

> cd openbabel-build && make install && cd ..

> ./compile_mychem

> cd mychem-build && make install && cd ..

> mysql -u root -p < mychem/src/mychemdb.sql

2.1.6. Create MySQL user for DiSCuS

You can do it via CLI or via PhpMyAdmin

2.1.7. Get DiSCuS code

> cd /var/www

> git clone https://github.com/mwojcikowski/discus.git discus

2.1.8. Open your instance of DiSCuS in web browser and proceed with installation.

> http://YOUR_HOSTNAME_OR_IP/discus/

2.1.9. DiSCuS is ready for use

Note: additional setup might be necessary for some plugins, f.e. you must get and copy your Xscore code, install Tripos Sybyl on host, etc.


jsme/release_notes.txt
 TODO: 
     - bug :D&D icon dissappears after resizing JSME window
     - scaling picture: still buggy for very small depiction

JSME_2013-09-29:
     - bug fix in index.html
JSME_2013-09-28:
     - More simple installation examples
     - index.html file
     - improved doc.html file
     - star option documented
     

JSME_2013-09-19:
    - New methods: addClickHandler(), addMouseOverHandler()

JSME_2013-09-17:
    - New methods:  getOffsetSize(), getOffsetWidth(), getOffsetHeight(), setWidth(), setHeight(),  isVisible(), setVisible()
    - improved javadoc API
    
JSME_2013-09-15:
    - new utility function: Utils.runAfterBrowserEventLoopReturns
    - setSize() method
    - call back function argument for replaceAllAppletsByJSME
    - JME: cache font and font metrics for the drawing area useful for many depictions
    - scaling picture bug fix: the 2px padding around the atom labels was not scaled

JSME_2013-09-08:
    - Easy migration JME->JSME
    - new documentation file: doc.html - still incomplete
    - new option: paste/nopaste to disable pasting MOL into the applet
    - query mode disabled by default
    - aromaticity SMARTS query bug fix

JSME_2013-08-04:
    -xbutton, noxbutton options work
    
JSME_2013-06-29:
    - For IE11, the clipboard management had to be adapted because of the unconventional user agent in IE11 (=Firefox)
    - Android and iPad do not show the drag out button because dragging out does not work on these systems
    - bug fix: undo & redo was not working for atom atom mapping
    - Note: the number of GWT permutations has increased to 30


JSME_2013-06-09:
    - atom move atom button is optional (turned off by default)
    - improved JSME_chemical_resolver_demo.html page


JSME_2013-05-30:
    - move atom button
    - java application exits when its window is closed, no more END button
    - new simpler HTML demo page
    
JSME_2013-05-01:
    - bug fix: IE has no console unless the debugger window is open
    - deleting explicit hydrogens when importing a structure from chemical resolver (demo page)

JSME_2013-04-09:
    - Custom paste menu and custom paste operation
    - non GPL licensed image for the drag and drop icon
    - local antialias for VML
    - line width for the molecular area
    - JavaDoc of the API
    - clear function added 

JSME_2013-04-01:
    - bug fix in JSObject

JSME_2013-03-30 :
    - Drag out of MOL/RXN and drop onto another applet or on to a file system window to save the file
    - DnD demo page
    - No more alert window if the function jsmeOnLoad is not defined - instead provides a message to the console
    - clear SDF stack as well when clicking on the clear icon
    - stereo bond cancel bug fix (bug shows up visible with undo/redo)
    - interactive atom marking demo page
    - multiple instances of the editor in the same HTML page work
    
JSME_2013-03-24 :
    - Drag and drop of files onto the applet

JSME_2013-03-21 :
    - undo bug fix
    - showInfo function


JSME_2013-03-19 :
    - antialias is on by default for SVG and off for VML
    - new help page
    - first public release

JSME_2013-03-18 :
    - spiro button
    - triple bond in menu is improved - looks better with browser with poor antialias


JSME_2013-03-16 :
    - rename JavaScriptApplet to JSApplet to avoid trademark problems with JavaScript
    - disable molecule scaling for IOS
    - removes blue marking about atoms and bonds after click enabled unless touch device


JSME_2013-03-14 :
    - increased touch radius for atoms and bonds - much better usability on iPad

JSME_2013-03-13 : not for redistribution
    - first DIV shield - still bugs
    - rubber banding box was not saved in the undo stack - fixed

JSME_2013-03-12 : not for redistribution
    - Rollback of all the Smiles changes, no more example of NotifyStructuralChangeJSfunction provided
    - improved rotation&scaling on iPad. Scaling is still buggy
JSME_2013-03-11 : not for redistribution
    - Smiles generation changes the atom ordering of the atoms in the editor, which is a problem for the NotifyStructuralChangeJSfunction. Fixed.
JSME_2013-03-10 : not for redistribution
    - Scaling of the molecule implemented for touch devices - still buggy. Problem with rotation, does not work on Android
    - NotifyStructuralChangeJSfunction - problem with smiles generation
JSME_2013-03-07
    - Marked atoms are not displayed anymore with a blue square after touch/mouse up
    - Touch device: rotation bug fixed (random jump move at the end of the rotation)
    - bug fix for CSS distribution
JSME_2013-03-04
    - Touch device: rotation works on iPad


JSME_2013-03-03
    - Touch device: touch and hold will show the copy&paste popup menu
    - Touch device: mouse events of drawing area were replaced by touch events in order to have a faster response on the iPad
    - Touch device: structure can be moved. It is smooth on a iPad 4, slow on a Samsung Note
    - Touch device: bond angle can be changed, multichain works
    - CSS pollution: all global CSS definitions were removed from all the included CSS files from GWT and gwt-mosaic
    - CSS global protection: better protection against global CSS definitions from the host page
    
    
    
Known bug: scaling and bond length rubber banding
    

common_jquery.js
/* Show loading overlay */
function loading(load) {
	load = typeof load == 'undefined' ? true : load
	if(load) {
		$('#loading').show();
	}
	else {
		$('#loading').hide();
	}
}
// bind keys to visualizer
$(document).keydown(function(e){
	if(Jmol) { 
		switch(e.which) {
//			case 37: // left

//			break;

			case 38: // up
			showPrevConformation();
			break;

//			case 39: // right

//			break;

			case 40: // down
			showNextConformation();
			break;

			default: return; // exit this handler for other keys
		}
		e.preventDefault();
	}
});

// query form vars
$('.query_delete_var').click(function() {
	if($(this).parent('div').parent('div').children('div.query_var').length > 1) {
		var container = $(this).parent('div').parent('div');
		$(this).parent('div').slideUp('fast',function(){
			$(this).remove();
			var first = container.children('div.query_var').eq(0).children('select[name^="var_logic"]').eq(0)
			first.css('visibility','hidden')
			first.children().removeAttr("selected");
			first.val("AND");
		});
		
	}
	else {
		$(this).parent('div').children('select[name^="var"]').each(function() { 
			$(this).removeAttr('selected');
			$(this).children().removeAttr("selected");
		});
		$(this).parent('div').children('input[name^="var"]').each(function() { 
			$(this).val('');
		});
	}
});
$('.query_add_var').click(function() {
	var new_var = $(this).parent().clone(true)
	new_var.hide().insertAfter($(this).parent()).slideDown('fast')
	new_var.children('select[name^="var_logic"]').eq(0).css('visibility','')
});
// query form targets
$('.query_delete_target').click(function() {
	if($('div.container_target').length > 1) {
		$(this).parent('div').parent('div').slideUp('fast',function(){ 
			$(this).remove();
			$('.container_target').first().map(function() {
				var first = $('select[name^="var_target_logic"]', this).first()
				first.css('visibility','hidden')
				first.children().removeAttr("selected");
			});
			// show "Any target"
			if($('div.container_target').length == 1) {
				$('select[name="var_target[]"] > option[value="-1"]').show();
			}
		 });
	}
	else {
		
		$(this).parent('div').parent('div').map(function() {
			$('select[name^="var_target"]', this).children().removeAttr("selected");
			// remove all but first query var and clean it's value
			$('div.query_var:not(:first)', this).slideUp('fast',function(){ 
				$(this).remove();
			});
			$('select[name^="var"]', this).each(function() { 
				$(this).removeAttr('selected');
				$(this).children().removeAttr("selected");
			});
			$('input[name^="var"]', this).each(function() { 
				$(this).val('');
			});
			// show "Any target"
			$('select[name="var_target[]"] > option[value="-1"]').show();
		});
		
	}
	
});
$('.query_add_target').click(function() {
	var new_target = $('div.container_target').last().clone(true);
	new_target.hide().insertAfter($('div.container_target').last()).slideDown('fast');
	$('select[name^="var_target_logic"]', new_target).eq(0).css('visibility','');
	// renumber vars
	var i = 0;
	$('div.container_var', $('div.container_target')).each(function() {
		$('select[name^="var"], input[name^="var"]', $(this)).each(function() { 
			$(this).attr('name', $(this).attr('name').replace(/\[(\d+)\]/g, '['+ i +']'))
		});
		i = i + 1;
	});
	// Hide "Any target" from view
	if($('div.container_target').length > 1) {
		$('select[name="var_target[]"] > option[value="-1"]').each(function() {
			$(this).hide()
			if($(this).attr('selected') || $(this).parent().val() == "-1") {
				$(this).removeAttr('selected');
				$(this).parent().map(function() {
					$(this).val($(this).children(':not([value=-1])').first().val());
				});
			}
		});
		$('select[name="var_target[]"]').trigger('change');
	}
});

$('select[name="var_target[]"]').change(function() {
	if($('div.container_target').length > 1) {
		$('select[name="sort_target"] > option:not([value=""])').each(function() {
			$(this).hide();
		});
		$('select[name="var_target[]"]').each(function() {
			$('select[name="sort_target"] > option[value="' + $(this).val() +'"]').show();
		});
		
	}
	else {
		$('select[name="sort_target"] > option').each(function() {
			$(this).show();
		});
	}
});

$('input:checkbox[name="disable_mol_grouping"]').change(function() {
	if($(this).is(":checked")) {
		$('select[name="sort_target"]').map(function() {
			$(this).prop('disabled', 'disabled')
			$(this).val("")
			$(this).children().removeAttr("selected");
		});
	}
	else {
		$('select[name="sort_target"]').prop('disabled', false);
	}
});
// rankscore targets
$('.rankscore_delete_target').click(function() {
	if($('ul.container_target').children('li').length > 1) {
		$(this).parent('li').slideUp('fast',function(){ 
			$(this).remove();
			// show "Any target"
			if($('ul.container_target').children('li').length == 1) {
				$('select[name^="target_id"] > option[value=""]', $('ul.container_target').children('li')).show();
			}
		 });
	}
	else {
		
		$(this).parent('li').map(function() {
			$('select[name^="target_id"]', $(this)).children().removeAttr("selected");
			// show "Any target"
			$('select[name^="target_id"] > option[value=""]', $(this)).show();
		});
		
	}
	
});
$('.rankscore_add_target').click(function() {
	var new_target = $('ul.container_target').children('li').last().clone(true);
	new_target.hide().insertAfter($('.container_target').children('li').last()).slideDown('fast');
	// Hide "Any target" from view
	if($('ul.container_target').children('li').length > 1) {
		$('select[name="target_id[]"] > option[value=""]', $('ul.container_target')).each(function() {
			$(this).hide()
			if($(this).attr('selected') || $(this).parent().val() == "") {
				$(this).removeAttr('selected');
				$(this).parent().map(function() {
					$(this).val($(this).children(':not([value=""])').first().val());
				});
			}
		});
		$('select[name="target_id[]"]', $('ul.container_target')).trigger('change');
	}
});


// data chart series
$('.delete-series').click(function() {
	if($('.series').length > 1) {
		$(this).parent().slideUp('fast',function(){ 
			$(this).remove();
		});
	}
});
$('#add-series').click(function() {
	$('#series-container').append($('.series').last().clone(true))
});

// notify on JSME structural change
function JSME_update() {
	$('#smiles').val(document.JSME.smiles());
}

// notify on smiles change
$('#smiles-clean').click(function() {
	convert_smi2sdf($('#smiles').val())
});

// convert SMILES using NCI REST
function convert_smi2sdf(smiles) {
	$.ajax({
		url: "openbabel_ajax.php",
		crossDomain: true,
		data: {
		  "smiles" : encodeURIComponent(smiles),
		  "output" : "sdf"
		},
		success: function( data ) {
			document.JSME.readMolFile(data);
		},
		error: function(jqXHR, textStatus, errorThrown) {
			("Ajax error: " + errorThrown);
		}
	 });
}

// draw JSME
$('#collapseOne').on('shown', function() {
	if(!document.JSME) {
		drawJsme()
	}
});

// invert selection button
$('input[name^="invert"]').click(function() {
	$('input[name="'+$(this).attr('name').replace('invert-', '')+'"]').each(function () {
		$(this).attr('checked', !$(this).attr('checked'));
	});
});
// select all button
$('input[name^="selall"]').click(function() {
	if($(this).is(':checkbox')) {
		var get_selection = $(this).prop('checked');
	}
	else {
		var get_selection = !$('input[name="'+$(this).attr('name').replace('selall-', '')+'"]').first().prop('checked');
	}
	$('input[name^="'+$(this).attr('name').replace('selall-', '')+'"]').each(function () {
		$(this).prop('checked', get_selection);
	});
});

// selection form downloads
$('input[type="checkbox"]', $('form[name="selection-form"]')).click(function() {
	if($('input:checked', $('form[name="selection-form"]')).length > 0) {
		$('ul.selected-download').css('display', '');
		$('ul.selected-download').parent('li').removeClass('disabled');
		$('ul.selected-subset').css('display', '');
		$('ul.selected-subset').parent('li').removeClass('disabled');
	}
	else {
		$('ul.selected-download').hide();
		$('ul.selected-download').parent('li').addClass('disabled');
		$('ul.selected-subset').hide();
		$('ul.selected-subset').parent('li').addClass('disabled');
	}
});

// highlight rows
$('table.molecules tr').hover(function() {
	$(this).prevUntil('tr:not(.'+$(this).attr('class')+')').last().children('td[rowspan]').each(function() {
		$(this).addClass("hovered");
	});
}, function () {
	$(this).prevUntil('tr:not(.'+$(this).attr('class')+')').last().children('td[rowspan]').each(function() {
		$(this).removeClass("hovered");
	});
});

// click on row to select
$('table.molecules tr').click(function() {
	if(event.target.type != 'checkbox' && event.target.type != 'button') {
		// click on self
		$(this).children().children('input:checkbox').first().each(function() { $(this).attr('checked', !$(this).attr('checked')) })
		// click on spanned
		$(this).prevUntil('tr:not(.'+$(this).attr('class')+')').last().children().children('input:checkbox').first().each(function() { $(this).attr('checked', !$(this).attr('checked')) })
	}
});

// resize molecules images on hover
$("img.mol").hover(function() {
	$(this).stop(true,true).delay(1000).animate({
		width: 300,
		height: 300,
	});
}, function() {
	$(this).stop(true,true).delay(1000).animate({
		width: 100,
		height: 100,
	});
});

// binding profile
$('div.binding_profile_residue').hover(function() {
	//$(this).animate( { "opacity" : .3 }, 50);
	if($('button.binding_profile_preview.btn-success').length > 0) {
		// check if is highlighted as interaction
		if($('#'+$(this).attr('id'), $('button.binding_profile_preview.btn-success').parent()).children('div.interaction').length > 0) {
			style = ['wireframe 50', 'label "%[group]%[resno]"'];
		}
		else {
			style = ['color cpk', 'wireframe 50', 'label "%[group]%[resno]"'];
		}
	}
	else {
		style = ['wireframe 50', 'label "%[group]%[resno]"'];
	}	
	$(this).stop(true,true).animate( { "opacity" : .3 }, 50).delay(500).queue(function() {
		if($(this).attr('id')) {
			script = ['select 1.1 resno = ' + $(this).attr('id'), style.join("; ")];
			Jmol.script(jmolApplet0,script.join("; "));
		}
	});
}, function() {
	if($('button.binding_profile_preview.btn-success').length > 0) {
		// check if is highlighted as interaction
		if($('#'+$(this).attr('id'), $('button.binding_profile_preview.btn-success').parent()).children('div.interaction').length > 0) {
			style = ['wireframe 30', 'label off'];
		}
		else {
			style = ['color darkblue', 'wireframe', 'label off'];
		}
	}
	else {
		style = ['wireframe', 'label off'];
	}
	if($(this).queue().length > 1) {
		$(this).stop(true,true).animate( { "opacity" : 1 }, 50);
	}
	else {
		$(this).stop(true,true).animate( { "opacity" : 1 }, 50).delay(100).queue(function() {
			if($(this).attr('id')) {
				script = ['select 1.1 resno = ' + $(this).attr('id'), style.join("; ")];
				Jmol.script(jmolApplet0,script.join("; "));
			}
		});
	}
});

$('button.binding_profile_preview').click(function() {
	if(!$(this).hasClass('btn-success') && $('button.binding_profile_preview.btn-success').length > 0) {
		$('button.binding_profile_preview.btn-success').trigger('click');
	}
	ires = [];
	$('div.binding_profile_residue', $(this).parent()).each(function() {
		if($(this).children('div.interaction').length > 0) {
			if($(this).attr('id')) {
				ires.push($(this).attr('id'));
			}
		}
	});
	script = ['select 1.1 and (resno = ' + ires.join(', resno = ') + ')'];
	if($(this).hasClass('btn-success')) {
		script.push('wireframe');
		script.push('color darkblue');
		$(this).removeClass('btn-success').removeClass('active');
	}
	else {
		script.push('color cpk');
		script.push('wireframe 30');
		//change color of rest of the protein
		script.push('select 1.1 and not selected; color darkblue');
		$(this).addClass('btn-success').addClass('active');
		
	}
	Jmol.script(jmolApplet0,script.join("; "));
	// check if any binding profile is active and return to CPK color scheme
	if($('button.binding_profile_preview.btn-success').length == 0) {
		Jmol.script(jmolApplet0,'select; color cpk;')
	}
	
});

$('input[name="crude-checkbox"]').click(function() {
	if($(this).is(":not(:checked)")) {
		$('div.crude').show();
	}
	else {
		$('div.crude').hide();
	}
});

// reset selection on acordion colapse
$('div.accordion-body').each(function () {
	$(this).on('hidden', function () {
		$('input:not(input[type=button],input[type=submit],button), select', this).each(function() {
			$(this).val('');
	  		$(this).removeAttr('selected');
	  	});
	});
});
